unliftio/src/UnliftIO/Async.hs

```diff
@@ -570,49 +611,87 @@ instance Applicative FlatApp where
 instance E.Exception ConcException
 
 -- | Simple difference list, for nicer types below
 type DList a = [a] -> [a]
 
+dlistConcat :: DList a -> DList a -> DList a
+dlistConcat = (.)
+{-# INLINE dlistConcat #-}
+
+dlistCons :: a -> DList a -> DList a
+dlistCons a as = dlistSingleton a `dlistConcat` as
+{-# INLINE dlistCons #-}
+
+dlistConcatAll :: [DList a] -> DList a
+dlistConcatAll = foldr (.) id
+{-# INLINE dlistConcatAll #-}
+
+dlistToList :: DList a -> [a]
+dlistToList = ($ [])
+{-# INLINE dlistToList #-}
+
+dlistSingleton :: a -> DList a
+dlistSingleton a = (a:)
+{-# INLINE dlistSingleton #-}
+
+dlistEmpty :: DList a
+dlistEmpty = id
+{-# INLINE dlistEmpty #-}
+
+
 -- | Turn a 'Conc' into a 'Flat'. Note that thanks to the ugliness of
 -- 'empty', this may fail, e.g. @flatten Empty@.
+--
 flatten :: forall m a. MonadUnliftIO m => Conc m a -> m (Flat a)
 flatten c0 = withRunInIO $ \run -> do
 
+  -- why not app?
   let both :: forall k. Conc m k -> IO (Flat k)
       both Empty = E.throwIO EmptyWithNoAlternative
       both (Action m) = pure $ FlatApp $ FlatAction $ run m
-      both (LiftA2 f a b) = do
-        a' <- both a
-        b' <- both b
-        pure $ FlatApp $ FlatLiftA2 f a' b'
-      both (Alt a b) = do
-        a' <- alt a
-        b' <- alt b
-        case a' $ b' [] of
-          [] -> E.throwIO EmptyWithNoAlternative
-          [x] -> pure $ FlatApp x
+      both (Lift cf ca) = do
+        f <- both cf
+        a <- both ca
+        pure $ FlatApp $ FlatLift f a
+      both (LiftA2 f ca cb) = do
+        a <- both ca
+        b <- both cb
+        pure $ FlatApp $ FlatLiftA2 f a b
+      both (Alt ca cb) = do
+        a <- alt ca
+        b <- alt cb
+        case dlistToList (a `dlistConcat` b) of
+          []    -> E.throwIO EmptyWithNoAlternative
+          [x]   -> pure $ FlatApp x
           x:y:z -> pure $ FlatAlt x y z
       both (Pure a) = pure $ FlatApp $ FlatPure a
 
       -- Returns a difference list for cheaper concatenation
       alt :: forall k. Conc m k -> IO (DList (FlatApp k))
-      alt Empty = pure id
-      alt (Alt a b) = do
-        a' <- alt a
-        b' <- alt b
-        pure $ a' . b'
-      alt (Action m) = pure (FlatAction (run m):)
-      alt (LiftA2 f a b) = do
-        a' <- both a
-        b' <- both b
-        pure (FlatLiftA2 f a' b':)
-      alt (Pure a) = pure (FlatPure a:)
+      alt Empty = pure dlistEmpty
+      alt (Lift cf ca) = do
+        f <- both cf
+        a <- both ca
+        pure (dlistSingleton $ FlatLift f a)
+      alt (Alt ca cb) = do
+        a <- alt ca
+        b <- alt cb
+        pure $ a `dlistConcat` b
+      alt (Action m) = pure (dlistSingleton $ FlatAction (run m))
+      alt (LiftA2 f ca cb) = do
+        a <- both ca
+        b <- both cb
+        pure (dlistSingleton $ FlatLiftA2 f a b)
+      alt (Pure a) = pure (dlistSingleton $ FlatPure a)
 
   both c0
 
+-- TODO: Write test that verifies that FlatAlt is actually Flat
```

Comment:
TODO needs to be implemented

Reply:
There is an inconvenience related to this; we are not exporting the types required to test this functionality (all `Flat` values are private to this module), should we go ahead and add an `Internals` module structure to the project? Wanted to check with you given there is nothing like that present right now, and I might guess this is on purpose.

Reply:
I'm OK with an `Internals` module here.